activitysim/abm/models/tour_mode_choice_reallocation.py

```diff
@@ -0,0 +1,653 @@
+    logsum_trips.index.name = "trip_id"
+
+    return logsum_trips
+
+
+def append_tour_leg_trip_mode_choice_logsums(tours):
+    """
+    Creates trip mode choice logsum column in tours table for each tour mode and leg
+
+    Parameters
+    ----------
+    tours : pd.DataFrame
+
+    Returns
+    -------
+    tours : pd.DataFrame
+        Adds two * n_modes logsum columns to each tour row, e.g. "logsum_DRIVE_outbound"
+    """
+    trips = inject.get_table("trips").to_frame()
+    trip_dir_mode_logsums = trips.pivot(
+        index="tour_id",
+        columns=["tour_mode", "outbound"],
+        values="trip_mode_choice_logsum",
+    )
+    new_cols = [
+        "_".join(["logsum", mode, "outbound" if outbound else "inbound"])
+        for mode, outbound in trip_dir_mode_logsums.columns
+    ]
+    trip_dir_mode_logsums.columns = new_cols
+    trip_dir_mode_logsums.reindex(tours.index)
+    tours = pd.merge(tours, trip_dir_mode_logsums, left_index=True, right_index=True)
+
+    return tours
+
+
+def get_trip_mc_logsums_for_all_modes(tours, segment_column_name, model_settings, trace_label):
+    """
+    Creates pseudo-trips from tours and runs trip mode choice to get logsums
+
+    Parameters
+    ----------
+    tours : pandas.DataFrame
+    segment_column_name : str
+        column in tours table used for segmenting model spec
+    model_settings : dict
+    trace_label : str
+
+    Returns
+    -------
+    tours : pd.DataFrame
+        Adds two * n_modes logsum columns to each tour row, e.g. "logsum_DRIVE_outbound"
+    """
+
+    # create pseudo-trips from tours for all tour modes
+    logsum_trips = create_logsum_trips(
+        tours, segment_column_name, model_settings, trace_label
+    )
+
+    # temporarily register trips in the pipeline
+    pipeline.replace_table("trips", logsum_trips)
+    tracing.register_traceable_table("trips", logsum_trips)
+    pipeline.get_rn_generator().add_channel("trips", logsum_trips)
+
+    # run trip mode choice on pseudo-trips. use orca instead of pipeline to
+    # execute the step because pipeline can only handle one open step at a time
+    orca.run(["trip_mode_choice"])
+
+    # add trip mode choice logsums as new cols in tours
+    tours = append_tour_leg_trip_mode_choice_logsums(tours)
+
+    # de-register logsum trips table
+    pipeline.get_rn_generator().drop_channel("trips")
+    tracing.deregister_traceable_table("trips")
+
+    return tours
+
+
+def code_household_auto_sufficiency(households):
+    """
+    Adds and codes household auto sufficiency based on household auto ownership and number of drivers in the household
+
+    Parameters
+    ----------
+    households : pd.Dataframe
+
+    Returns
+    -------
+    households : pd.Dataframe
+        Household table with auto sufficiency column added
+
+    """
+    households['auto_sufficiency'] = 'suff'
+    households.loc[
+        households.auto_ownership.between(0, households.num_drivers, 'neither'), 'auto_sufficiency'] = 'insuff'
```

Comment:
@TianYangLin Can we double check how auto sufficiency is defined in the rest of the model? I think in other parts, it's based on number vehicles vs number of workers. What we did makes sense, just want to make note of how its defined in the rest of the model.

Reply:
See reply to comment above on line 190